modules/rostests/apitests/ntdll/LdrLoadDll.c

```diff
@@ -0,0 +1,107 @@
+    IN PULONG DllCharacteristics OPTIONAL,
+    IN PUNICODE_STRING DllName,
+    OUT PVOID *BaseAddress
+);
+
+*/
+
+#define StartSeh() \
+    ExceptionStatus = STATUS_SUCCESS; \
+    _SEH2_TRY
+
+#define EndSeh(ExpectedStatus) \
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER) { \
+        ExceptionStatus = _SEH2_GetExceptionCode(); \
+    } _SEH2_END; \
+    ok(ExceptionStatus == ExpectedStatus, "Exception %lx, expected %lx\n", \
+       ExceptionStatus, ExpectedStatus)
+
+#define InvalidPointer ((PVOID)0x0123456789ABCDEFULL)
+
+START_TEST(LdrLoadDll)
+{
+    NTSTATUS Status, ExceptionStatus;
+    UNICODE_STRING DllName;
+    PVOID BaseAddress, BaseAddress2;
+    WCHAR szWinDir[MAX_PATH], szSysDir[MAX_PATH], szPath[MAX_PATH];
+
+    RtlInitEmptyUnicodeString(&DllName, NULL, 0);
+    GetWindowsDirectoryW(szWinDir, _countof(szWinDir));
+    GetSystemDirectoryW(szSysDir, _countof(szSysDir));
+
+    StartSeh() { LdrLoadDll(NULL, NULL, &DllName, NULL);          } EndSeh(STATUS_ACCESS_VIOLATION);
+    StartSeh() { LdrLoadDll(NULL, NULL, &DllName, &BaseAddress);  } EndSeh(STATUS_ACCESS_VIOLATION);
+    StartSeh() { LdrLoadDll(L"", NULL, &DllName, NULL);           } EndSeh(STATUS_ACCESS_VIOLATION);
+    StartSeh() { LdrLoadDll(L"", NULL, &DllName, &BaseAddress);   } EndSeh(STATUS_ACCESS_VIOLATION);
+
+    RtlInitUnicodeString(&DllName, L"advapi32.dll");
+    StartSeh()
+    {
+        Status = LdrLoadDll(NULL, NULL, &DllName, NULL);
+        if (NT_SUCCESS(Status)) LdrUnloadDll(BaseAddress);
+    } EndSeh(STATUS_ACCESS_VIOLATION);
+
+    RtlInitUnicodeString(&DllName, L"advapi32.dll");
+    StartSeh()
+    {
+        BaseAddress = InvalidPointer;
+        Status = LdrLoadDll(NULL, NULL, &DllName, &BaseAddress);
+        ok(Status == STATUS_SUCCESS, "Status = %lx\n", Status);
+        ok(BaseAddress != NULL && BaseAddress != InvalidPointer, "BaseAddress = %p\n", BaseAddress);
+        if (NT_SUCCESS(Status))
+        {
+            BaseAddress2 = InvalidPointer;
+            Status = LdrLoadDll(NULL, NULL, &DllName, &BaseAddress2);
+            ok(Status == STATUS_SUCCESS, "Status = %lx\n", Status);
+            ok(BaseAddress2 == BaseAddress, "BaseAddress2 = %p, expected %p\n", BaseAddress2, BaseAddress);
+            LdrUnloadDll(BaseAddress);
+        }
+    } EndSeh(STATUS_SUCCESS);
+
+    RtlInitUnicodeString(&DllName, L"advapi32.dll");
+    StartSeh()
+    {
+        Status = LdrLoadDll(L"\\SystemRoot\\System32", NULL, &DllName, NULL);
+    } EndSeh(STATUS_ACCESS_VIOLATION);
+
+    RtlStringCchPrintfW(szPath, _countof(szPath), L"%s\\advapi32", szSysDir);
+
+    RtlInitUnicodeString(&DllName, szPath);
+    StartSeh()
+    {
+        Status = LdrLoadDll(szWinDir, NULL, &DllName, &BaseAddress);
+        ok(Status == STATUS_SUCCESS, "Status = %lx\n", Status);
+        if (NT_SUCCESS(Status)) LdrUnloadDll(BaseAddress);
+    } EndSeh(STATUS_SUCCESS);
+
+    RtlStringCchPrintfW(szPath, _countof(szPath), L"%s/advapi32", szSysDir);
```

Comment:
What happens if the szSysDir contains only forward slashes as well?

Reply:
Still successful.